File: DSA/Graph/Code/ConnectedComponents.py
# Connected Components in Python

from collections import deque

class ConnectedComponents:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adjacency_matrix = [[0] * vertices for _ in range(vertices)]
    
    def add_edge(self, u, v):
        self.adjacency_matrix[u][v] = 1
        self.adjacency_matrix[v][u] = 1
    
    def find_components(self):
        visited = [False] * self.vertices
        component = 0
        
        print("Connected Components:")
        for i in range(self.vertices):
            if not visited[i]:
                print(f"Component {component + 1}: ", end="")
                component += 1
                
                queue = deque([i])
                visited[i] = True
                
                while queue:
                    u = queue.popleft()
                    print(u, end=" ")
                    
                    for v in range(self.vertices):
                        if self.adjacency_matrix[u][v] and not visited[v]:
                            visited[v] = True
                            queue.append(v)
                
                print()

# Test
print("=== Connected Components (Python) ===\n")

graph = ConnectedComponents(6)

graph.add_edge(0, 1)
graph.add_edge(0, 2)
graph.add_edge(3, 4)

print("Graph (6 vertices):")
print("Component 1: 0-1, 0-2")
print("Component 2: 3-4")
print("Isolated: 5\n")

graph.find_components()

print("\n=== Algorithm ===")
print("1. Initialize all vertices as unvisited")
print("2. For each unvisited vertex:")
print("   a. Start BFS/DFS from that vertex")
print("   b. Mark all reachable vertices as visited")
print("   c. This forms one connected component")
print("3. Repeat until all vertices visited\n")

print("=== Complexity ===")
print("Time Complexity:  O(V + E)")
print("Space Complexity: O(V)")

File: DSA/Graph/Code/ConnectedComponents.cs
// Connected Components in C#

using System;
using System.Collections.Generic;

class ConnectedComponents {
    private int vertices;
    private int[,] adjacencyMatrix;
    
    public ConnectedComponents(int v) {
        vertices = v;
        adjacencyMatrix = new int[v, v];
    }
    
    public void AddEdge(int u, int v) {
        adjacencyMatrix[u, v] = 1;
        adjacencyMatrix[v, u] = 1;
    }
    
    public void FindComponents() {
        bool[] visited = new bool[vertices];
        int component = 0;
        
        Console.WriteLine("Connected Components:");
        for (int i = 0; i < vertices; i++) {
            if (!visited[i]) {
                Console.Write("Component " + (++component) + ": ");
                
                Queue<int> q = new Queue<int>();
                q.Enqueue(i);
                visited[i] = true;
                
                while (q.Count > 0) {
                    int u = q.Dequeue();
                    Console.Write(u + " ");
                    
                    for (int v = 0; v < vertices; v++) {
                        if (adjacencyMatrix[u, v] == 1 && !visited[v]) {
                            visited[v] = true;
                            q.Enqueue(v);
                        }
                    }
                }
                Console.WriteLine();
            }
        }
    }
    
    static void Main() {
        Console.WriteLine("=== Connected Components (C#) ===\n");
        
        ConnectedComponents graph = new ConnectedComponents(6);
        
        graph.AddEdge(0, 1);
        graph.AddEdge(0, 2);
        graph.AddEdge(3, 4);
        
        Console.WriteLine("Graph (6 vertices):");
        Console.WriteLine("Component 1: 0-1, 0-2");
        Console.WriteLine("Component 2: 3-4");
        Console.WriteLine("Isolated: 5\n");
        
        graph.FindComponents();
        
        Console.WriteLine("\n=== Algorithm ===");
        Console.WriteLine("1. Initialize all vertices as unvisited");
        Console.WriteLine("2. For each unvisited vertex:");
        Console.WriteLine("   a. Start BFS/DFS from that vertex");
        Console.WriteLine("   b. Mark all reachable vertices as visited");
        Console.WriteLine("   c. This forms one connected component");
        Console.WriteLine("3. Repeat until all vertices visited\n");
        
        Console.WriteLine("=== Complexity ===");
        Console.WriteLine("Time Complexity:  O(V + E)");
        Console.WriteLine("Space Complexity: O(V)");
    }
}
